src/nats.rs

```diff
@@ -152,6 +152,107 @@ pub async fn ensure_stream(
         .map_err(|e| anyhow::anyhow!("{e:?}"))
 }
 
+pub async fn ensure_limits_stream(
+    context: &Context,
+    name: String,
+    subjects: Vec<String>,
+    description: Option<String>,
+) -> Result<Stream> {
+    let stream_config = StreamConfig {
+        name: name.clone(),
+        description,
+        num_replicas: 1,
+        retention: async_nats::jetstream::stream::RetentionPolicy::Limits,
+        subjects,
+        max_age: DEFAULT_EXPIRY_TIME,
+        storage: async_nats::jetstream::stream::StorageType::File,
+        allow_rollup: false,
+        ..Default::default()
+    };
+
+    if let Ok(stream) = context.get_stream(&name).await {
+        // For now, we only check if the subjects are the same in order to make sure that
+        // newer versions of wadm adjust subjects appropriately. In the case that developers
+        // want to alter the storage or replicas of a stream, for example,
+        // we don't want to override that configuration.
+        if stream.cached_info().config.subjects == stream_config.subjects {
+            return Ok(stream);
+        } else {
+            warn!("Found stream {name} with different configuration, deleting and recreating");
+            context.delete_stream(name).await?;
+        }
+    }
+
+    context
+        .get_or_create_stream(stream_config)
+        .await
+        .map_err(|e| anyhow::anyhow!("{e:?}"))
+}
+
+pub async fn ensure_event_consumer_stream(
+    context: &Context,
+    name: String,
+    subject: String,
+    streams: Vec<&Stream>,
+    description: Option<String>,
+) -> Result<Stream> {
+    let mut sources: Vec<Source> = vec![];
+    for source_stream in streams {
+        let mut subject_transforms: Vec<SubjectTransform> = vec![];
+
+        for source_stream_subject in &source_stream.cached_info().config.subjects {
+            // If we have a multi-tenant stream subject, we need to replace
+            // the second wildcard since the first one represents the account id,
+            // otherwise replace the first one:
+            //
+            // multi-tenant:  <account-id>.<subject>.evt.<lattice-id>.<event-type>
+            // single-tenant: <subject>.evt.<lattice-id>.<event-type>
+            let wildcard = match source_stream_subject.starts_with('*') {
+                true => "{{wildcard(2)}}",
+                false => "{{wildcard(1)}}",
+            };
+            let destination = subject.replacen('*', wildcard, 1);
+            subject_transforms.push(SubjectTransform {
+                source: source_stream_subject.clone(),
+                destination,
+            });
+        }
```

Comment:
Nit: Consider reducing at least this inner for loop to use an `iter` and `map` instead of a for loop. If you're feeling brave, you could also turn the outer loop into an iter as well

Reply:
it's funny you say that, because I actually had it implemented that way originally, but it got a little unwieldy to read, so I decided to go with a much simpler (admittedly more verbose) approach.

Happy to bring back the iter approach if we generally prefer that

Reply:
It is just a nit, so no requirement here

Reply:
Addressed this in the latest push.